### PR TITLE
Update dummy primitive generation rules

### PR DESCRIPTION
**Pull request by safern**
Fixes: https://github.com/dotnet/arcade/issues/3891

cc: @ahsonkhan @ericstj @stephentoub

### CODE REVIEW

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs [hunk 1]
```diff
@@ -168,33 +168,36 @@ public override void Visit(ITypeDefinition parentType, IEnumerable<IFieldDefinit
...
-                    var genericTypedFields = excludedFields.Where(f => f.Type.UnWrap().IsGenericParameter());
+                    // Struct layout should follow these rules:
+                    // 1. If there is a reference type field in the struct or within the fields' type closure,
+                    //    it should emit a reference type and a value type dummy field.
+                    //    - The reference type dummy field is needed in order to inform the compiler to block
+                    //      taking pointers to this struct because the GC will not track updating those references.
+                    //    - The value type dummy field is needed in order for the compiler to error correctly on definite assignment checks in all scenarios. dotnet/roslyn#30194
+                    // 2. If there are no reference type fields, but there are value type fields in the struct field closure,
+                    //    and at least one of these fields is a nonempty struct, then we should emit a value type dummy field.
+                    // 3. If the type is generic, then for every type parameter of the type, if there are any private
```
Comment (ahsonkhan) [inline src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs]: nit: Add the comment back about validating struct layout cycles to 3.

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs [hunk 2]
```diff
@@ -168,33 +168,36 @@ public override void Visit(ITypeDefinition parentType, IEnumerable<IFieldDefinit
...
-
-                    bool hasValueTypePrivateField = excludedFields.Any(f => !f.Type.IsOrContainsReferenceType());
-
-                    if (hasValueTypePrivateField)
+                    else
                     {
-                        IFieldDefinition fieldType = DummyFieldWriterHelper(parentType, excludedFields, parentType.PlatformType.SystemInt32, "_dummyPrimitive");
-                        newFields.Add(fieldType);
+                        bool hasEmptyValueTypePrivateField = excludedFields.Any(f => f.Type.IsNonEmptyStruct());
+                        if (hasEmptyValueTypePrivateField)
```
Comment (ahsonkhan) [inline src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs]: Should this be `!hasEmptyValueTypePrivateField`? If not, we should invert the name.

File: src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs [hunk 3]
```diff
@@ -149,6 +151,50 @@ public static bool IsOrContainsReferenceType(this ITypeReference type)
             return false;
         }
 
+        public static bool IsNonEmptyStruct(this ITypeReference type)
```
Comment (ahsonkhan) [inline src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs]: nit: rename as follows `IsOrContainsNonEmptyStruct`?

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs (hunk 1 above)
Comment (safern) [reply src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs]: Good catch.

File: src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs (hunk 2 above)
Comment (safern) [reply src/Microsoft.Cci.Extensions/Writers/CSharp/CSharpWriter.cs]: No, actually the name should be, `hasNonEmptyStructPrivateField`

File: src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs (hunk 3 above)
Comment (safern) [reply src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs]: Correct.

**Comment by ahsonkhan** [comment]
Question: In general (probably outside the scope of this PR), how do we go about testing GenAPI and tools like these within arcade? Would this benefit from unit testing to verify correctness?

**Comment by safern** [comment]
Yes currently I do manual testing myself, but yes, we have plans on adding tests (there is one issue, I can’t find now cause I’m in the phone), I’m planning on adding these tests soon.

**Comment by safern** [comment]
@ericstj would you mind taking a look?

File: src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs [hunk 4]
```diff
@@ -149,6 +151,54 @@ public static bool IsOrContainsReferenceType(this ITypeReference type)
...
+
+                    // If we're not in the root of the tree, it means we found a non-empty struct.
+                    return true;
+                }
+
+                foreach (var field in resolvedType.Fields.Where(f => !f.IsStatic))
+                {
+                    if (visited.Contains(field.Type))
+                    {
+                        // if we find a visited type in the fields tree, it means we have a cycle, which means the struct is non-empty.
```
Comment (ericstj) [inline src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs]: Could you have a cycle of empty types?

Comment (safern) [reply src/Microsoft.Cci.Extensions/Extensions/CSharp/CSharpCciExtensions.cs]: Updated to a better approach, thanks for taking a look.

**Comment by safern** [comment]
@JunTaoLuo @dougbu FYI this will cause the DARC PR to fail on aspnet/extensions, I ran the code check step locally and there are some structs that are missing a dummy primitive field because the rules where wrong in previous implementation.

**Comment by dougbu** [comment]
> there are some structs that are missing a dummy primitive field because the rules where wrong in previous implementation

@safern is the fix to rebuild Extensions' ref/ assemblies in the dependency update PR that brings the newer Arcade? If not, what action should we take?

**Comment by safern** [comment]
> @safern is the fix to rebuild Extensions' ref/ assemblies in the dependency update PR that brings the newer Arcade? If not, what action should we take?

Yes, that would be the fix, just wanted to give you a heads up.

**Pull request merged**